### PR TITLE
#779 Support for new lowest_version_diff property

### PR DESCRIPTION
**Pull request by witcher112**

### CODE REVIEW

File: Assets/PatchKit Patcher/Scripts/AppUpdater/AppUpdaterStrategyResolver.cs [hunk 1]
```diff
@@ -45,89 +47,118 @@ public StrategyType GetFallbackStrategy(StrategyType strategyType)
...
-                        "Installed version is older than the latest version. Checking whether cost of updating with diff is lower than cost of updating with content...");
-
-                    if (context.Configuration.CheckConsistencyBeforeDiffUpdate)
-                    {
-                        DebugLogger.Log("Checking consitency before allowing diff update...");
+                        int lowestVersionWithDiffId = GetLowestVersionWithDiffId(context);
+                        _logger.LogTrace("lowestVersionWithDiffId = " + lowestVersionWithDiffId);
 
-                        if (!IsVersionIntegral(context))
+                        if (installedVersionId + 1 < lowestVersionWithDiffId)
```
Comment (Dzejkop) [inline Assets/PatchKit Patcher/Scripts/AppUpdater/AppUpdaterStrategyResolver.cs]: I have some issues here:
1. The next log says `"Diffs for updating from installed to latest version are not available. Using content strategy."`, which, while technically true, does not seem exactly equivalent to what this condition is checking. Shouldn't the log look more like `"The installed version doesn't support diff upadtes. Using content strategy."`?

2. It took me a few seconds to understand the `+ 1` bit. A method along the lines of `DoesVersionSupportDiffUpdates(installedVersionId)` would be more readable.

File: Assets/PatchKit Patcher/Scripts/AppUpdater/AppUpdaterStrategyResolver.cs [hunk 2]
```diff
@@ -45,89 +47,118 @@ public StrategyType GetFallbackStrategy(StrategyType strategyType)
...
-                    DebugLogger.Log(string.Format("Cost of updating with diff equals {0}.", diffCost));
-                    DebugLogger.Log(string.Format("Cost of updating with content equals {0}.", contentCost));
+                        _logger.LogDebug(
+                            "Checking whether cost of updating with diff is lower than cost of updating with content...");
 
-                    if (diffCost < contentCost)
-                    {
-                        DebugLogger.Log(
-                            "Cost of updating with diff is lower than cost of updating with content. Using diff strategy.");
+                        long contentSize = GetLatestVersionContentSize(context);
```
Comment (Dzejkop) [inline Assets/PatchKit Patcher/Scripts/AppUpdater/AppUpdaterStrategyResolver.cs]: Why are content size and diff size types mismatched? Shouldn't they both be `ulong`?

File: Assets/PatchKit Patcher/Scripts/AppUpdater/AppUpdaterStrategyResolver.cs [hunk 3]
```diff
@@ -138,36 +169,35 @@ private bool IsVersionIntegral(AppUpdaterContext context)
...
+
+        private static long GetLatestVersionContentSize(AppUpdaterContext context)
         {
             int latestVersionId = context.App.GetLatestVersionId();
             var contentSummary = context.App.RemoteMetaData.GetContentSummary(latestVersionId);
             return contentSummary.Size;
         }
 
-        private ulong GetDiffCost(AppUpdaterContext context)
+        private static ulong GetSumLatestVersionDiffSize(AppUpdaterContext context)
```
Comment (Dzejkop) [inline Assets/PatchKit Patcher/Scripts/AppUpdater/AppUpdaterStrategyResolver.cs]: Might be just me but this is just funny 😆 

As in "Come get sum Latest Version Diff Size"

Perhaps `GetLatestVersionDiffSizeSum` ?

File: Assets/PatchKit Patcher/Scripts/AppUpdater/AppUpdaterStrategyResolver.cs (hunk 1 above)
Comment (Dzejkop) [reply Assets/PatchKit Patcher/Scripts/AppUpdater/AppUpdaterStrategyResolver.cs]: Ad. 2

**Comment by genail** [comment]
@witcher112 Is there a possibility to write an integration test for that?

**Comment by witcher112** [comment]
@genail unfortunately it's not possible right now.

**Pull request merged**